fix kbaseTabs behaving incorrectly when a deletable tab is deleted [PTV-1845]

eapearson opened the pull request:
# Description of PR purpose/changes

The venerable "kbaseTabs" widget is behaving incorrectly in the kbaseExpressionConditionsetBaseWIdget.

Specifically, it is not showing the previous tab when a dynamic tab is closed, as is done when clicking on a cluster.

While fixing the kbaseTabs widget, and expanding the tests to cover additional use case, I discovered at least one additional latent bug (or at least unexpected behavior), a failure to activate the first tab if no tabs are set explicitly as active. 

In addition, there were several features in this already rather complex widget that are not used anywhere in the codebase, and have been removed. They may have been used in the past, but I suspect they were "nice to have" features that were simply added for future utility. However, retaining them would require adding tests to cover them, yet yield no benefit as they are unused.

Finally, tests were expanded to cover common use cases, including the primary issue fixed by this PR.

# Jira Ticket / Issue #

Related Jira ticket: https://kbase-jira.atlassian.net/browse/DATAUP-X
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions

New tests were added to test/unit/spec/widgets/kbaseTabs-spec.js to cover the use case of closing a dynamically added tab.

To evaluate this manually, spin up a narrative from this PR, create a narrative, add a FeatureClusters object, and open the viewer. 

Unfortunately it is not possible to manually evaluate all usage scenarios with the Narrative, as they are not all utilized or demonstrable. To cover those cases I used a small gallery feature which allows embedding a widget in a static html page in order to demonstrate all usage scenarios. That code is not included in this PR. If "storybook" is added to the Narrative, such demonstrations of UI functionality can be included with PRs.

- [x] Tests pass locally and in GitHub Actions
  - see comment below; ui unit tests pass in isolation, but inevitably fail when all run in sequence.
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
  - ui unit tests appear unaltered after these changes; the tests are generally non-deterministic - in isolation all tests
    pass, but when all run together, there are inevitably failures.
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook
- [-] (Python) I have run Black and Flake8 on changed Python code manually or with a git precommit hook
- [-] Any dependent changes have been merged and published in downstream modules

# Updating Version and Release Notes (if applicable)

- [-] [Version has been bumped](https://semver.org/) for each release
- [x] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
+    'use strict';
 
-*/
+    const $el = (tagName) => {
```
A better way to create jQuery wrapped bare elements

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
         options: {
-            tabPosition: 'top',
             canDelete: false,
-            borderColor: 'lightgray',
```
unused, removed

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
+        version: '1.1.0',
 
         options: {
-            tabPosition: 'top',
```
unused, removed

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
 
-        version: '1.0.0',
-
-        _accessors: ['tabsHeight'],
```
unused, removed

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -64,167 +69,218 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
 
         appendUI: function ($elem, tabs) {
-            if (tabs == undefined) {
+            if (typeof tabs === 'undefined') {
```
safer test for undefinedness.

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -64,167 +69,218 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
 
         addTab: function (tab) {
-            if (tab.canDelete == undefined) {
+            if (typeof tab.canDelete === 'undefined') {
```
safer test for undefinedness.

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -64,167 +69,218 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
             if (tab.content) {
-                $tab.append(tab.content);
-                $tab.hasContent = true;
+                $tabPanel.append(tab.content);
```
renamed for clarity

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -64,167 +69,218 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
+                $tabPanel.hasContent = true;
             }
 
-            if (this.options.border) {
```
removing border feature, as unused.

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -241,47 +297,72 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
         removeTab: function (tabName) {
             const $tab = this.data('tabs')[tabName];
             const $nav = this.data('nav')[tabName];
+            const deleteNav = () => {
```
extracted into functions as they are re-used, and also used separately.

eapearson commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -241,47 +297,72 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
-            const $deleteModal = new kbaseDeletePrompt($('<div></div>'), {
+            const $deleteModal = new $DeletePrompt($el('div'), {
                 name: tabName,
-                callback: this.deleteTabCallback(tabName),
```
simplified as there was no need to have an anonymous callback function call a method of such brevity

ialarmedalien commented on docs/developer/local-docker.md:
```diff
@@ -50,6 +59,97 @@ This will launch the integration test  runner against the CI environment, using
+
+Unit tests require a running Narrative server in order to access Jupyter javascript files which are only available when the server is built and running.
+
+As a server, it can be accessed within a container is well, and for those who prefer Docker-based workflows, is quite convenient.
```
> within a container _as_ well

ialarmedalien replied:
(I think that's what it's supposed to be, anyway!)

eapearson replied:
If you can believe it, I'm back on this PR, nearly a year to the day 🙀 
And, yes, is -> as.

ialarmedalien commented on docs/developer/local-docker.md:
````diff
@@ -50,6 +59,97 @@ This will launch the integration test  runner against the CI environment, using
+Then, to run tests without authentication.
+
+```shell
+nvm use 16
````
is there something earlier in the guide about using nvm for node management?

eapearson replied:
Nope. You know, this document has been sitting in the repo for many years with very little change. It was basically to memorialize how I develop the Narrative with docker, rather than running it on the bare host with Python. It isn't a "blessed" Narrative workflow - as far as I know, no-one else uses it. Well, no-one ever commented on it before!

So I'm happy to make changes to it and will, but that is why it is fairly bare-bones.

ialarmedalien replied:
Wow, that comment was from over a year ago. I was a different person then!

ialarmedalien commented on docs/developer/local-docker.md:
````diff
@@ -50,6 +59,97 @@ This will launch the integration test  runner against the CI environment, using
+   13 total       13 passed      0 failed      0 skipped 
+```
+
+To focus tests, you can modify `karma.local.conf.js`. Here we have focussed on the tests that
+are directly contained within `test/unit/spec/widgets/*`, which at the moment, is just `kbaseTabs-spec.js`.
````
You might want to add comments about how to do this using the native jasmine API -- edit the test files to use `fdescribe` or `fit` in place of `describe` or `it` respectively (and `xdescribe` and `xit` to skip tests).

eapearson replied:
Good point. I just wanted to point out this other technique for focusing/excluding at the directory level.

And I do appreciate that you actually read this humble little doc and suggested improvements.

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
     var $tabs =  new kbaseTabs($('#tabs'), {
-            tabPosition : 'bottom', //or left or right or top. Defaults to 'top'
-            canDelete : true,       //whether or not the tab can be removed. Defaults to false.
+            canDelete : true,       // boolean; whether a tab which does not otherwise specify this option
+                                    //   will have a close button and can be removed. Defaults to false, overridden
+                                    //   if set on a tab.
```
are these lines Prettier compliant? I can't remember if it reformats commented text or not. I ask because the lines look quite long and are wrapping a bit oddly in the GitHub display

eapearson replied:
Prettier does not format comments, other than jsdoc ones. Yes, these are very wide, have been that way for probably over a decade.

Nevertheless, I extracted the code, moved the comments around fixed and modernized the code.

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
-                    showContentCallback: function // if you don't want to show the content right away, add a callback method that returns the content...
+                    show : true,                                    // boolean; If true the tab is shown by default. If not specified, the first tab is shown.
+                    showContentCallback: function                   // function returning jquery "content" (see above); renders tab pane content to enable
+                                                                    //   "lazy" rendering of a tab pane when the corresponding tab is
```
...when the corresponding tab is...?

the suspense!!

eapearson replied:
I know, right? shown? Well, in bootstrap 3 parlance, tabs are shown and hidden

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -2,54 +2,59 @@
     var $tabs =  new kbaseTabs($('#tabs'), {
-            tabPosition : 'bottom', //or left or right or top. Defaults to 'top'
-            canDelete : true,       //whether or not the tab can be removed. Defaults to false.
+            canDelete : true,       // boolean; whether a tab which does not otherwise specify this option
```
do you need to add `confirmDelete` here too?

eapearson replied:
I'm adding it in, although my guess is that either the docs were created before the confirmDelete option was added, or this specific example has canDelete set to false, in which case confirmDelete is not necessary.

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -64,167 +69,218 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
-                    $.proxy(function (idx, tab) {
-                        this.addTab(tab);
+            if (tabs && tabs.length > 0) {
+                // defaults to first tab being active if none are set as active by default.
```
does a check need to be added to ensure that only one tab has `show` set to true? What happens if more than one tab is set to `show: true`?

eapearson replied:
Good point, troublemaker. Do we need to protect from programmer error here?
Adding this.

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -64,167 +69,218 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
+
+        renderDeleteTabButton: function (tab) {
+            return $el('button')
+                .addClass('btn btn-default btn-xs')
+                .append($el('span').append('&#x2716;'))
+                .css('padding', '0px')
+                .css('width', '12px')
+                .css('height', '12px')
+                .css('margin-left', '10px')
+                .css('font-size', '10px')
+                .css('margin-bottom', '3px')
+                .css('outline', 'none')
+                .css('border', '0')
```
Can this hard-coded css be replaced by an extra class, e.g. `btn-delete` or `btn-xs-delete` or something similar?

eapearson replied:
replaced with .kbaseTabs-delete-button

ialarmedalien commented on test/unit/spec/widgets/kbaseTabs-spec.js:
```diff
@@ -1,20 +1,357 @@
+     * <link type="text/css" rel="stylesheet" href="/narrative/static/style/style.min.css">
+     * (or wherever the bootstrap styles appear) as it contains the bootstrap styles.
+     *
+     * The reliance upon the 'bootstrap' dependency above does not incorporate bootstratp styles.
```
typo: `bootstratp`

eapearson replied:
fixed

eapearson commented:
I've addressed the issues above, run through the grist mill of merging with conflicts, linting and formatting battles.
Added the infamous "gallery" back in, as it is truly the only way to see the different states of tabs, it is difficult to convey this in words, and is essential in developing it, as there is no interface in the narrative which uses all tab features.

ialarmedalien commented on docs/developer/local-docker.md:
```diff
@@ -1,6 +1,17 @@
+- nodejs 16
+- nvm
+
+> Sorry, this is just a quick guide, no instructions provided for installing these.
```
If people don't know how to install them, they shouldn't even be reading this file!

eapearson replied:
That is true, but sometimes I do memorialize installation instructions, too, in the hope it imparts some experience to novices. I know many people still just go to, say, the nodejs site and download whatever they see. E.g. they are on macos and they don't use macports or homebrew (yet.)

ialarmedalien commented on docs/developer/local-docker.md:
````diff
@@ -50,6 +70,118 @@ This will launch the integration test  runner against the CI environment, using
+```javascript
+          xit('should render with minimal options', () => {
+            const kbaseTabs = makeKBaseTabs({
+```
````
Great docs. I appreciate that you write documentation as these kinds of things are so easy to forget as soon as you move on to another project.

eapearson replied:
no kidding.

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbasePrompt2.js:
```diff
@@ -0,0 +1,200 @@
+    At this time, it is also a proper AMD module, whilst the older one is not (but still
+    works, as it is loaded globally).
+
+    It also uses viewer of the 
```
don't leave us in suspense!

eapearson replied:
and now we'll never know.

ialarmedalien replied:
😢

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -1,55 +1,82 @@
+        tabs : [
+            {
+                // string; label of the tab
+                tab : 'T1',
```
would you be able to use something that's more obviously a tab title (e.g. "Tab Title 1") - whenever I see something like `'t1'`, `'t2'`, etc., it makes me wonder if it's a variable.

eapearson replied:
Not my original documentation, but I'll try to improve it.

ialarmedalien replied:
Thanks, it is much clearer now!

ialarmedalien commented on kbase-extension/static/kbase/js/widgets/kbaseTabs.js:
```diff
@@ -241,45 +338,73 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
+                const [$nextTab, $nextPanel] = (() => {
+                    if ($nav.next('li').length) {
+                        return [$nav.next('li'), $tab.next('[role="tabpanel"]')];
+                    } else if ($nav.prev('li').length) {
```
No need for the `else` - can just do 
```js
if (condition_one) {
   return whatever;
}
if (condition_two) {
   return something_else;
}
return [null, null];
```

eapearson replied:
Good catch, and I am a big fan of unraveling sequences of conditional branches into a sequence of if statements which just return, just got lazy here, I think, as I was transforming the original code and just adapted it. Although, to be fair, it is only stylistic as everyone understands a sequence of if-else. Still, changed.

ialarmedalien commented on test/unit/testUtil.js:
```diff
@@ -67,14 +68,14 @@ define('testUtil', [
-     *      {function} domStateFunction function that returns true when the watched event happens
-     *      {function} executeFirst     a function to execute after setting up the observer
-     * @param {string} selector to identify the element being watched for
+     * @param {Object} args.config configuration for the MutationObserver
+     * @param {string} args.documentElement  DOM element to watch
+     * @param {function} args.domStateFunction function that returns true when the
+     * watched event happens
+     * @param {function} args.executeFirst     a function to execute after setting up
+     * the observer
```
those aren't separate params - they are keys/values in the `args` object

eapearson replied:
Right, that is how objects are documented in jsdoc, I believe. It is a bit weird.

ialarmedalien replied:
Huh. Weird!

eapearson replied:
Although another way of doing this is to create type definitions with @typedef and object properties described with @property, and then just refer to this type in the @param. Unless these types are used by a typechecker, and they serve primarily for documentation (and IDE support if available), I guess best would be whatever is most readable?

ialarmedalien commented on test/unit/testUtil.js:
```diff
@@ -154,10 +158,77 @@ define('testUtil', [
+     * look for it within the element; a function will accept a string to look in, and
+     * will return a boolean indicating whether it was found.
+     * @param {Object} options (optional) several options to control behavior
+     * @param {number} options.timeout (optional) duration, in milliseconds, for which
+     * to look for the string; defaults to 1000ms
+     * @param {number} options.interval (optional) the duration, in milliseconds,
+     * between attempts to find the string; defaults to 100ms
+     * @param {string} options.textName (optional) the a name for the text being looked
+     * for, used in the test failure message; defaults to "text".
```
this is a confusing way to describe the keys of the options object (IMO)

eapearson replied:
Isn't that how js doc objects are documented?

ialarmedalien commented on kbase-extension/static/kbase/gallery/README.md:
```diff
@@ -0,0 +1,44 @@
+# Gallery
```
I'm dubious about adding this gallery functionality, especially since I can see it getting neglected and/or not used by anyone but you. I don't think I would bother to add stuff to it. @briehl your thoughts?

eapearson replied:
I realize this, but honestly I added this because I don't know how anyone could review the changes without actually using them live in a browser, and this is the most direct way I could think of. I've had this code around for over a year, and every time I return to it I realize how useful it is. It has the benefits of storybook in that you can create any number of variations of a widget. (I think the last time I tried to get it in, that was BIll's comment - that storybook would eventually be added and take on the same responsibility). But this doesn't add any new dependencies, it is just a good ol static web app.

eapearson commented on kbase-extension/static/kbase/js/widgets/kbasePrompt2.js:
```diff
@@ -0,0 +1,200 @@
+            // into the DOM, in case you were wondering how that magic happens.
+            this.$dialogModal = this.createDialogModal();
+            this.$dialogModal.on('hidden.bs.modal', function () {
+                // Zap the model.
```
This was actually in response to a problem that arose in testing, I wish I remembered what the issue was, exactly. By default bootstrap just keeps the modal hanging around but hidden, which can cause issues (don't remember exactly what the problem was).  Completely removing the modal when done has no discernable performance impact, as far as I could tell, and for those unfamiliar with the design of bootstrap, probably more expected.

ialarmedalien replied:
I'm fairly sure I've run into some issues with modals hanging around after they have been dismissed by the user.

ialarmedalien commented on kbase-extension/static/kbase/gallery/common.css:
```diff
@@ -0,0 +1,20 @@
+.Demo {
```
if we're going to have a gallery, may as well make it official and integrate this into the common CSS file.

eapearson replied:
Okay, good idea; css has been moved to `_gallery.scss` in `partials`. Was trying to keep the gallery decoupled (it depends on the Narrative source, but Narrative source doesn't know know about it.)

eapearson commented:
I get 2-4 tests failing locally, if set concurrency is set to 1. Without that setting, many tests involving a dialog fail with a timeout - with about 20 failed tests in total.
